File: langs/csharp/src/VMx/Localization/NullLocalizer.cs
namespace VMx.Localization;

/// <summary>
/// Null-object variant per ADR-0017. Both overloads return the input key
/// unchanged. See spec/17-localization.md.
/// </summary>
public sealed class NullLocalizer : ILocalizer
{
    /// <summary>Shared singleton.</summary>
    public static NullLocalizer Instance { get; } = new();

    private NullLocalizer() { }

    /// <inheritdoc/>
    public string Localize(string key) => key;

    /// <inheritdoc/>
    public string Localize(string key, IEnumerable<object?> args) => key;
}

File: langs/python/src/vmx/localization/null_localizer.py
"""NullLocalizer — null-object variant per ADR-0017."""

from __future__ import annotations

from collections.abc import Iterable


class NullLocalizer:
    """Returns the input key unchanged (no localization, no formatting)."""

    def localize(self, key: str, args: Iterable[object] | None = None) -> str:
        return key


NULL_LOCALIZER: NullLocalizer = NullLocalizer()
"""Shared singleton."""
